src/services/admin/api-key.ts

```diff
@@ -0,0 +1,169 @@
+import type { Repository } from 'typeorm';
+import { decodeJWT } from 'did-jwt';
+import { randomBytes } from 'crypto';
+import { Connection } from '../../database/connection/connection.js';
+
+import * as dotenv from 'dotenv';
+import type { CustomerEntity } from '../../database/entities/customer.entity.js';
+import { APIKeyEntity } from '../../database/entities/api.key.entity.js';
+import type { UserEntity } from '../../database/entities/user.entity.js';
+import { SecretBox } from '@veramo/kms-local';
+import { API_KEY_LENGTH, API_KEY_PREFIX, API_KEY_EXPIRATION } from '../../types/constants.js';
+import pkg from 'js-sha3';
+import { v4 } from 'uuid';
+import type { APIServiceOptions } from '../../types/portal.js';
+dotenv.config();
+
+const { sha3_512 } = pkg;
+
+export class APIKeyService {
+	public apiKeyRepository: Repository<APIKeyEntity>;
+	private secretBox: SecretBox;
+
+	public static instance = new APIKeyService();
+
+	constructor() {
+		this.apiKeyRepository = Connection.instance.dbConnection.getRepository(APIKeyEntity);
+		this.secretBox = new SecretBox(process.env.EXTERNAL_DB_ENCRYPTION_KEY);
+	}
+
+	// ToDo: Maybe we also need to store not the API key but the hash of it?
+	// But in that case the API key will be shown once and then it will be lost.
```

Comment:
No reason to do so per se.

If handled properly, the user will save the secret on their end, along with our handling storing it safely encrypted, decrypting it in transit, if necessary.

If we're looking to compare the provided API key in headers, then we _should_ compute the HMAC SHA-256 / SHA-512 from the plain-text value **and** a user-specific unique identifier, such as the Logto short unique id or other.

The reconstructed value _must_ be calculated deterministically, in which case we _should_ store the same result in the DB on creation following the same process.

Reply:
The same can be achieved with bcrypt > 11 rounds.